lib/cusolver/cusolverRF.jl

```diff
@@ -0,0 +1,429 @@
+            spH,
+            n, nnzA, desca,
+            h_rowsA, h_colsA, h_Qreorder,
+        )
+    elseif ordering == :METIS
+        cusolverSpXcsrmetisndHost(
+            spH,
+            n, nnzA, desca,
+            h_rowsA, h_colsA, C_NULL, h_Qreorder,
+        )
+    elseif ordering == :RCM
+        cusolverSpXcsrsymrcmHost(
+            spH,
+            n, nnzA, desca,
+            h_rowsA, h_colsA, h_Qreorder,
+        )
+    end
+
+    h_mapBfromA = zeros(Cint, nnzA)
+    @inbounds for i in 1:nnzA
+        h_mapBfromA[i] = i # identity matrix
+    end
+
+    # Compute permutation in two steps
+    size_perm = Ref{Csize_t}(0)
+    cusolverSpXcsrperm_bufferSizeHost(
+        spH,
+        m, n, nnzA, desca,
+        h_rowsB, h_colsB, h_Qreorder, h_Qreorder,
+        size_perm,
+    )
+
+    buffer_cpu = Base.Libc.malloc(size_perm[] * sizeof(Cint))
+    cusolverSpXcsrpermHost(
+        spH,
+        m, n, nnzA, desca,
+        h_rowsB, h_colsB, h_Qreorder, h_Qreorder, h_mapBfromA,
+        buffer_cpu,
+    )
+
+    # Apply permutation
+    h_valsB = h_valsA[h_mapBfromA]
+
+    # LU Factorization
+    info = Ref{CUSOLVER.csrqrInfo_t}()
+    cusolverSpCreateCsrluInfoHost(info)
+
+    cusolverSpXcsrluAnalysisHost(
+        spH,
+        m, nnzA, desca,
+        h_rowsB, h_colsB, info[],
+    )
+
+    size_internal = Ref{Cint}(0)
+    size_lu = Ref{Cint}(0)
+    cusolverSpDcsrluBufferInfoHost(
+        spH,
+        n, nnzA, desca,
+        h_valsB, h_rowsB, h_colsB,
+        info[],
+        size_internal, size_lu
+    )
+
+    n_bytes = size_lu[] * sizeof(Cint)
+    buffer_lu = Base.Libc.malloc(n_bytes)
+    pivot_threshold = 1.0
+
+    cusolverSpDcsrluFactorHost(
+        spH, n, nnzA, desca,
+        h_valsB, h_rowsB, h_colsB,
+        info[], pivot_threshold,
+        buffer_lu,
+    )
+
+    # Check singularity
+    if check
+        singularity = Ref{Cint}(0)
+        cusolverSpDcsrluZeroPivotHost(
+            spH, info[], tol, singularity,
+        )
+
+        # Check that the matrix is nonsingular
+        if singularity[] >= 0
+            SingularException(singularity[])
+        end
+    end
+
+    # Get size of L and U
+    pnnzU = Ref{Cint}(0)
+    pnnzL = Ref{Cint}(0)
+    cusolverSpXcsrluNnzHost(
+        spH,
+        pnnzL, pnnzU, info[],
+    )
+
+    nnzL = pnnzL[]
+    nnzU = pnnzU[]
+
+    # Retrieve L and U matrices
+    h_Plu = zeros(Cint, m)
+    h_Qlu = zeros(Cint, n)
+
+    h_valsL = zeros(nnzL)
+    h_rowsL = zeros(Cint, m+1)
+    h_colsL = zeros(Cint, nnzL)
+
+    h_valsU = zeros(nnzU)
+    h_rowsU = zeros(Cint, m+1)
+    h_colsU = zeros(Cint, nnzU)
+
+    # Extract
+    cusolverSpDcsrluExtractHost(
+        spH,
+        h_Plu, h_Qlu,
+        desca,
+        h_valsL, h_rowsL, h_colsL,
+        desca,
+        h_valsU, h_rowsU, h_colsU,
+        info[],
+        buffer_lu,
+    )
+
+    h_P = h_Qreorder[h_Plu .+ 1]
+    h_Q = h_Qreorder[h_Qlu .+ 1]
+
+    if buffer_lu != C_NULL
```

Comment:
`nothing` as a sentinel would be better. But how can `buffer_lu` be NULL here? You just allocated it?

Reply:
I added a check as I am used to do that in C when wrapping `free` ... At least, as we are not using `Base.Libc.malloc` anymore, we can get rid of these lines